src/python/pants/backend/python/tasks2/pytest_run.py

```diff
@@ -366,15 +308,109 @@ def pex_run(args):
+          removed = 0
+          for i, item in enumerate(list(items)):
+            if i % {nshards} != {shard}:
+              del items[i - removed]
+              removed += 1
+          reporter = config.pluginmanager.getplugin('terminalreporter')
+          reporter.write_line('Only executing {{}} of {{}} total tests in shard {shard} of '
+                              '{nshards}'.format(total_count - removed, total_count),
+                              bold=True, invert=True, yellow=True)
+        """.format(shard=sharder.shard, nshards=sharder.nshards))
+    except Sharder.InvalidShardSpec as e:
+      raise self.InvalidShardSpecification(e)
+
+  def _get_conftest_content(self, sources_map):
+    # A conftest hook to modify the console output, replacing the chroot-based
+    # source paths with the source-tree based ones, which are more readable to the end user.
+    # Note that python stringifies a dict to its source representation, so we can use sources_map
+    # as a format argument directly.
+    console_output_conftest_content = dedent("""
+
+      ### GENERATED BY PANTS ###
+
+      import pytest
+
+      # Map from source path relative to chroot -> source path relative to buildroot.
+      _SOURCES_MAP = {}
+
+      @pytest.hookimpl(hookwrapper=True)
+      def pytest_collect_file(path, parent):
+        outcome = yield
+        res = outcome.get_result()
+        for module in res:
+          # nodeid is used for sorting and display, so it's safe to modify it.
+          # Note that module._nodeid is a lazily computed field, while module.nodeid
+          # is the property that computes and caches it.
+          module._nodeid = _SOURCES_MAP.get(module.nodeid, module.nodeid)
+    """.format(sources_map))
+    # Add in the sharding conftest, if any.
+    shard_conftest_content = self._get_shard_conftest_content()
+    return (console_output_conftest_content + shard_conftest_content).encode('utf8')
+
+  @contextmanager
+  def _conftest(self, sources_map):
+    """Creates a conftest.py to customize our pytest run."""
+
+    conftest_content = self._get_conftest_content(sources_map)
+    # We must create our conftest.py in the root of the source chroot, so that its hooks
+    # run while loading the test files in that chroot.  We take care not to stomp on an
+    # existing conftest.py from the source code.  We temporarily append to it instead.
+    conftest = os.path.join(
+      self.context.products.get_data(GatherSources.PYTHON_SOURCES).path(), 'conftest.py')
+    conftest_orig = None
+    # If there's an existing conftest.py, overwrite it to ensure it contains exactly
+    # any original user content, plus our temporary additions.
+    if os.path.exists(conftest):
+      # First, ensure that we've saved the original conftest.py.
+      conftest_orig = '{}.orig'.format(conftest)
+      if not os.path.exists(conftest_orig):
+        # We only copy if the orig file isn't present, to ensure that the conftest.py we're
```

Comment:
This comment doesn't seem accurate... it could definitely be the one from a previous failed run.